docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,255 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Redefine Container Runtime Interface
+
+The umbrella issue: [#22964](https://issues.k8s.io/22964)
+
+## Motivation
+
+Kubelet employs a declarative pod-level interface, which acts as the sole
+integration point for container runtimes (e.g., `docker` and `rkt`). The
+high-level, declarative interface has caused higher integration and maintenance
+cost, and also slowed down feature velocity for the following reasons.
+  1. **Not every container runtime supports the concept of pods natively**.
+     When integrating with Kubernetes, a significant amount of work needs to
+     go into implementing a shim of significant size to support all pod
+     features. This also adds maintenance overhead (e.g., `docker`).
+  2. **High-level interface discourages code sharing and reuse among runtimes**.
+     E.g, each runtime today implements an all-encompassing `SyncPod()`
+     function, with the Pod Spec as the input argument. The runtime implements
+     logic to determine how to achieve the desired state based on the current
+     status, (re-)starts pods/containers and manages lifecycle hooks
+     accordingly.
+  3. **Pod Spec is susceptible to inconsistent interpretations**. E.g., `rkt`
+     does not support container-level operations and assume immutable pods.
+  4. **Pod Spec is evolving rapidly**. New features are being added constantly.
```

Comment:
> Just wanted to point out first that lifecyles of the containers in a pod are not tied together even in today's k8s. kubelet restarts (or more correctly, create and start a new container) a container in a pod for you. E.g., if you have a side-car container for logging and crashed, there is no reason to kill your working application container and restart the entire pod for the crashed side-car container.

It depends on your vision of pods, I'd like pods to be small and considered as a composite container, in which case I'd split the pod if it doesn't make sense to restart it as a whole when a component fails. Whereas your example is the opposite.

> I don't think container-level operations make pod less application-oriented. IMHO, we are trying to make applications run well on k8s, and they are the requirements today.

It's not exactly 'alloc', but my point is that it's becoming similar to 'alloc'.

> As for `alloc`, it is still different because k8s doesn't allow adding containers in a pod yet (based on my very limited Borg knowledge). However, if it turns out that allowing updates are what users want and how the pod evolves, what's wrong with doing that?


Fair point, but IMHO sometimes you need to steer users towards good solutions, Golang is the biggest example of this philosophy.